### PR TITLE
Release of updated version for HP using apikey and apisecret 

### PR DESCRIPTION
**Pull request by Framoh76**
Release of first version of the Updated HP plugin. 
This version uses the ApiKey and ApiSecret that can be obtained creating an account on the hp warranty api website: https://developers.hp.com/css-enroll
After enrollment process, one should wait for an approval from HP for getting the access and create an APP that will provide you a key/secret pair.

The account provides the access to 100 requests/day. WIth the possibility to ask HP to upgrade the account to more than 100/day.

### CODE REVIEW

File: inc/config.class.php [hunk 1]
```diff
@@ -105,6 +105,14 @@ function preconfig($type) {
             if ($type == self::DELL) {
                $this->fields["supplier_key"] = "123456789";
             }
+			/* MODIF HP */
```
Comment (ddurieux) [inline inc/config.class.php]: Remove the comment, not need it

File: inc/config.class.php [hunk 2]
```diff
@@ -105,6 +105,14 @@ function preconfig($type) {
...
+            if ($type == self::HP) {
+			   $apiKey = '123456789990900AAAAAA';
```
Comment (ddurieux) [inline inc/config.class.php]: Wrong indentation

File: inc/config.class.php [hunk 3]
```diff
@@ -342,6 +350,21 @@ function showForm ($ID, $options = []) {
       } else {
          echo "<input type='hidden' name='warranty_duration' value='0'>\n";
       }
+	// MODIF HP
```
Comment (ddurieux) [inline inc/config.class.php]: Remove the comment

File: inc/config.class.php [hunk 4]
```diff
@@ -342,6 +350,21 @@ function showForm ($ID, $options = []) {
...
+      if ($this->fields["name"] == self::HP) {
+         echo "<tr>";
+         echo "<td class='tab_bg_2 center' colspan='2'>".__('Manufacturer API key', 'manufacturersimports')."</td>";
+         echo "<td class='tab_bg_2 left' colspan='2'>";
+         echo "<input type='text' name='supplier_key' size='32' value='".$this->fields["supplier_key"]."'>";
```
Comment (ddurieux) [inline inc/config.class.php]: Increase the size to 100 to be sure in the future not have problem if key more long

File: inc/config.class.php [hunk 5]
```diff
@@ -342,6 +350,21 @@ function showForm ($ID, $options = []) {
...
+         echo "</td>";
+         echo "</tr>";
+         echo "<tr>";
+         echo "<td class='tab_bg_2 center' colspan='2'>".__('Manufacturer API Secret', 'manufacturersimports')."</td>";
+         echo "<td class='tab_bg_2 left' colspan='2'>";
+         echo "<input type='text' name='supplier_secret' size='32' value='".$this->fields["supplier_secret"]."'>";
```
Comment (ddurieux) [inline inc/config.class.php]: idem

File: inc/hp.class.php [hunk 6]
```diff
@@ -44,143 +44,93 @@ function showDocTitle($output_type, $header_num) {
...
 
    /**
     * @see PluginManufacturersimportsManufacturer::getSupplierInfo()
     */
-   function getSupplierInfo($compSerial = null, $otherSerial = null, $key = null, $supplierUrl = null) {
+   function getSupplierInfo($compSerial = null, $otherSerial = null, $key = null, $supplierUrl = null, $secret = null) {
       $info["name"]          = PluginManufacturersimportsConfig::HP;
-      $info["supplier_url"] = "http://h20565.www2.hpe.com/hpsc/wc/public/find";
-
+	  $info["supplier_url "] = "https://css.api.hp.com";
```
Comment (ddurieux) [inline inc/hp.class.php]: fix the indent (use spaces instead tab)

File: inc/hp.class.php [hunk 7]
```diff
@@ -44,143 +44,93 @@ function showDocTitle($output_type, $header_num) {
...
+		foreach( $exp as $e => $val) {
+			$elt = explode(':', $val);
+			if( in_array('"endDate"', $elt)) {
+				$endDate = str_replace('"','', $elt[1]);
+				return $endDate; // EndDate found
+			}
+		}
+		return false;	// EndDate not found!!!  
+	}
+/* NEW method: */
```
Comment (ddurieux) [inline inc/hp.class.php]: Remove this comment

File: inc/postimport.class.php [hunk 8]
```diff
@@ -111,6 +111,29 @@ static function cURLData($options) {
...
          curl_setopt($ch, CURLOPT_POSTFIELDS, $post);
+
+		// ADDED FOR HP curl_setopt($ch, CURLOPT_HTTPHEADER, array('Accept: application/json'));
+		if( $options['suppliername'] == PluginManufacturersimportsConfig::HP) {
+			curl_setopt($ch, CURLOPT_HTTPHEADER, array('Accept: application/json'));
+			$token = $options["token"];
+			
+			if( $token != '') {
+				/*		Example: $postHP = '[{"sn": "CND3210W9M","pn": "D5H49AV"}]'; */
+				$postHP = '[{';
```
Comment (ddurieux) [inline inc/postimport.class.php]: Prefer use:
```
$postHP = [];
foreach( $options['post'] as $key => $value) {
   $postHP[$key] = $value;
}
and after use in _curl_setopt_ `json_encode($postHP)` 
```

File: inc/preimport.class.php [hunk 9]
```diff
@@ -162,7 +162,7 @@ static function getSupplierPost($suppliername, $compSerial, $otherserial = null)
       if (!empty($suppliername)) {
          $supplierclass = "PluginManufacturersimports" . $suppliername;
          $supplier      = new $supplierclass();
-         $infos         = $supplier->getSupplierInfo($compSerial, $otherserial);
+         $infos         = $supplier->getSupplierInfo($compSerial, $otherserial, $supplierKey, $supplierUrl, $supplierSecret);
```
Comment (ddurieux) [inline inc/preimport.class.php]: The variables `$supplierKey, $supplierUrl, $supplierSecret` are not defined, so I think there is a problem

File: inc/postimport.class.php [hunk 10]
```diff
@@ -379,8 +401,8 @@ static function seePostImport($type, $ID, $fromsupplier, $fromwarranty, $configI
          }
          echo "<a href='" . $link . "?id=" . $ID . "'>" . $line["name"] . $dID . "</a><br>" . $otherSerial . "</td>";
 
-         $url  = PluginManufacturersimportsPreImport::selectSupplier($suppliername, $supplierUrl, $compSerial, $otherSerial, $supplierkey);
-         $post = PluginManufacturersimportsPreImport::getSupplierPost($suppliername, $compSerial, $otherSerial);
+         $url  = PluginManufacturersimportsPreImport::selectSupplier($suppliername, $supplierUrl, $compSerial, $otherSerial, $supplierkey, $supplierSecret);
```
Comment (ddurieux) [inline inc/postimport.class.php]: This new argument not defined in the function `selectSupplier`

File: inc/postimport.class.php [hunk 11]
```diff
@@ -379,8 +401,8 @@ static function seePostImport($type, $ID, $fromsupplier, $fromwarranty, $configI
...
+         $post = PluginManufacturersimportsPreImport::getSupplierPost($suppliername, $compSerial, $otherSerial, $supplierkey, $supplierSecret); // HP ADD key,secret
```
Comment (ddurieux) [inline inc/postimport.class.php]: These new arguments not defined in the function `getSupplierPost`

File: inc/postimport.class.php (hunk 8 above)
Comment (Framoh76) [reply inc/postimport.class.php]: I've tried that, but doesn't work...

**Comment by ddurieux** [comment]
Thanks for the fixes, I will test it this afternoon

**Comment by ddurieux** [comment]
When create HP config, we have the error:
```
  *** PHP Notice(8): Undefined index: supplier_url
  Backtrace :
  ...s/manufacturersimports/inc/config.class.php:104 
  ...s/manufacturersimports/inc/config.class.php:267 PluginManufacturersimportsConfig->preconfig()
  inc/commonglpi.class.php:472                       PluginManufacturersimportsConfig->showForm()
  ajax/common.tabs.php:96                            CommonGLPI::displayStandardTab()
```
and so the API URL not filled.

When I import I have many error and have import KO:
```
2018-06-07 23:36:36 [2@ddurieux]
  *** PHP Notice(8): Undefined index: supplier_secret
  Backtrace :
  ...nufacturersimports/inc/postimport.class.php:371 
  ...nufacturersimports/inc/postimport.class.php:214 PluginManufacturersimportsPostImport::seePostImport()
  .../manufacturersimports/ajax/massiveaction.php:48 PluginManufacturersimportsPostImport::massiveimport()
2018-06-07 23:36:36 [2@ddurieux]
  *** PHP Notice(8): Undefined variable: supplierUrl
  Backtrace :
  ...anufacturersimports/inc/preimport.class.php:169 
  ...nufacturersimports/inc/postimport.class.php:404 PluginManufacturersimportsPreImport::getSupplierPost()
  ...nufacturersimports/inc/postimport.class.php:214 PluginManufacturersimportsPostImport::seePostImport()
  .../manufacturersimports/ajax/massiveaction.php:48 PluginManufacturersimportsPostImport::massiveimport()
2018-06-07 23:36:36 [2@ddurieux]
  *** PHP Notice(8): Undefined index: token
  Backtrace :
  ...nufacturersimports/inc/postimport.class.php:118 
  ...nufacturersimports/inc/postimport.class.php:481 PluginManufacturersimportsPostImport::cURLData()
  ...nufacturersimports/inc/postimport.class.php:422 PluginManufacturersimportsPostImport::saveImport()
  ...nufacturersimports/inc/postimport.class.php:214 PluginManufacturersimportsPostImport::seePostImport()
  .../manufacturersimports/ajax/massiveaction.php:48 PluginManufacturersimportsPostImport::massiveimport()
2018-06-07 23:36:37 [2@ddurieux]
  *** PHP Notice(8): Undefined property: stdClass::$access_token
  Backtrace :
  ...nufacturersimports/inc/postimport.class.php:485 
  ...nufacturersimports/inc/postimport.class.php:422 PluginManufacturersimportsPostImport::saveImport()
  ...nufacturersimports/inc/postimport.class.php:214 PluginManufacturersimportsPostImport::seePostImport()
  .../manufacturersimports/ajax/massiveaction.php:48 PluginManufacturersimportsPostImport::massiveimport()

```

**Comment by mohierf** [comment]
Import successfull for my computer on my configuration:
image


The problems:
 - create an `empty-2.0.2.sql` that creates a full table from scratch
 - it looks like the *preconfig* function is raising some php errors

Update the *hook.php* with:
```
function plugin_manufacturersimports_install() {
   global $DB;

   include_once(GLPI_ROOT . "/plugins/manufacturersimports/inc/profile.class.php");
   include_once(GLPI_ROOT . "/plugins/manufacturersimports/inc/config.class.php");

   $migration = new Migration("2.0.2");
   $update    = false;

   //Root of SQL files for DB installation or upgrade
   $sql_root = GLPI_ROOT . "/plugins/manufacturersimports/sql/";

   if (!$DB->tableExists("glpi_plugin_manufacturersimports_configs")) {

      $DB->runFile($sql_root . "/empty-2.0.2.sql");

```

**Comment by Framoh76** [comment]
I've fixed the Preimport issue and also add an empty-2.0.2.sql file for one installing that plugin from scratch.
Don't forget to get your own key/secret couple and to enter them into the plugin configuration form.

**Comment by mohierf** [comment]
LGTM. what do you think about it @ddurieux ?

**Comment by amma35** [comment]
Missing : 
Update to 2.0.2 in hook.php  
The indentation of the code is not good (3 spaces instead of tabulation)

**Comment by Framoh76** [comment]
@amma35 fixes done.

**Comment by amma35** [comment]
PHP errors stated by @ddurieux are not corrected

**Comment by mohierf** [comment]
No news from anyone to merge this  PR and release a version ?

**Pull request closed without merging**